src/main/java/teammates/ui/webapi/action/GetCoursesAction.java

```diff
@@ -17,17 +24,73 @@ protected AuthType getMinAuthLevel() {
 
     @Override
     public void checkSpecificAccessControl() {
-        gateKeeper.verifyInstructorPrivileges(logic.getAccount(userInfo.getId()));
+        String entityType = getNonNullRequestParamValue(Const.ParamsNames.ENTITY_TYPE);
+
+        if (!((entityType.equals(Const.EntityType.STUDENT) && userInfo.isStudent)
+                || (entityType.equals(Const.EntityType.INSTRUCTOR) && userInfo.isInstructor))) {
+            throw new UnauthorizedAccessException("Current account cannot access to courses of request entity type");
+        }
     }
 
     @Override
     public ActionResult execute() {
-        List<CourseAttributes> courses =
-                logic.getCoursesForInstructor(logic.getInstructorsForGoogleId(userInfo.getId(), true));
+        String entityType = getNonNullRequestParamValue(Const.ParamsNames.ENTITY_TYPE);
+        switch (entityType) {
+        case Const.EntityType.STUDENT:
+            return getStudentCourses();
+        case Const.EntityType.INSTRUCTOR:
+            return getInstructorCourses();
+        default:
+            return new JsonResult("Error: invalid entity type", HttpStatus.SC_BAD_REQUEST);
+
+        }
+    }
+
+    private ActionResult getStudentCourses() {
+        List<CourseAttributes> courses = logic.getCoursesForStudentAccount(userInfo.id);
+        return new JsonResult(new CoursesData(courses));
+    }
+
+    private ActionResult getInstructorCourses() {
+        String courseStatus = getNonNullRequestParamValue(Const.ParamsNames.COURSE_STATUS);
+        List<CourseAttributes> courses;
+        switch (courseStatus) {
+        case Const.CourseStatus.ACTIVE:
+            courses = getActiveCourse();
+            break;
+        case Const.CourseStatus.ARCHIVED:
+            courses = getArchivedCourse();
+            break;
+        case Const.CourseStatus.SOFT_DELETED:
+            courses = getSoftDeletedCourse();
+            break;
+        default:
+            return new JsonResult("Error: invalid course status", HttpStatus.SC_BAD_REQUEST);
+        }
 
-        courses.sort(Comparator.comparing(CourseAttributes::getId));
+        CourseAttributes.sortById(courses);
+        return new JsonResult(new CoursesData(courses));
 
-        return new JsonResult(new CourseInfo.CoursesResponse(courses));
     }
 
+    private List<CourseAttributes> getActiveCourse() {
+        return logic.getCoursesForInstructor(userInfo.id, true);
+    }
+
+    private List<CourseAttributes> getArchivedCourse() {
+
+        List<InstructorAttributes> allInstructors = logic.getInstructorsForGoogleId(userInfo.id, false);
+        List<InstructorAttributes> archivedInstructors = new ArrayList<>();
```

Comment:
If you are comfortable with lambda, it can be written in a more elegant way with filter. 

`allInstructors.stream().filter(instructor -> instructor.isArchived).collect(Collectors.toList());`

Reply:
Well, I seldom use it because I think it may make the code difficult to be understood by other developers